File: memory-gateway/app/context/compiler.py
"""Context compiler: fixed-token budget compilation, knapsack selection, and version snapshotting."""

from __future__ import annotations

import json
import logging
from dataclasses import dataclass, field
from typing import Any

from sqlmodel import select

from app.config import get_settings
from app.context.assembler import assemble_context_messages
from app.context.selector import (
    _extract_keywords,
    score_canonical_item,
    score_message_item,
    select_items_for_budget,
)
from app.context.tokens import estimate_messages_tokens, estimate_tokens
from app.memory.compressor import compress_tool_message
from app.memory.ids import normalize_message
from app.memory.state import latest_context_version, list_memory_items
from app.models.memory import MemoryStatus
from app.providers.memory_ai import MemoryAIAdapter
from app.storage.db import session_scope
from app.storage.models import ContextVersion, utcnow

logger = logging.getLogger(__name__)


@dataclass
class CompileResult:
    messages: list[dict[str, Any]]
    total_tokens: int
    context_version: int | None = None
    canonical_items_used: int = 0
    selected_count: int = 0
    budget: int = 8000


def _persist_context_snapshot(
    conversation_id: str,
    compiled_messages: list[dict[str, Any]],
    *,
    budget: int,
    total_tokens: int,
    canonical_count: int,
    source_message_ids: list[str],
) -> int | None:
    """Persist compiled context version snapshot into SQLite."""
    try:
        with session_scope() as session:
            next_ver = latest_context_version(session, conversation_id) + 1
            state_data = {
                "budget": budget,
                "total_tokens": total_tokens,
                "message_count": len(compiled_messages),
                "canonical_items_count": canonical_count,
            }
            row = ContextVersion(
                conversation_id=conversation_id,
                version=next_ver,
                state_json=json.dumps(state_data, ensure_ascii=False),
                source_message_ids_json=json.dumps(source_message_ids, ensure_ascii=False),
                created_at=utcnow(),
            )
            session.add(row)
            session.flush()
            return next_ver
    except Exception as exc:
        logger.warning(
            "Failed to persist context version snapshot for %s: %s",
            conversation_id,
            exc,
        )
        return None


async def compile_context(
    messages: list[dict[str, Any]],
    conversation_id: str | None = None,
    *,
    budget: int | None = None,
    memory_ai: MemoryAIAdapter | None = None,
    persist_snapshot: bool = True,
) -> CompileResult:
    """
    Compile fixed-budget context for upstream AI model.

    Invariants respected:
    - Never mutates main AI response
    - Fails open to original messages on error
    - No naive head/tail truncation
    - High-value decisions and constraints prioritized
    - Recent context preserved for coherence
    """
    settings = get_settings()
    target_budget = budget or settings.context_budget

    if not messages:
        return CompileResult(
            messages=[],
            total_tokens=0,
            budget=target_budget,
        )

    try:
        # Find latest user message to derive query keywords for relevance scoring
        latest_user_text = ""
        for m in reversed(messages):
            if m.get("role") == "user":
                latest_user_text = str(m.get("content") or "")
                break
        query_keywords = _extract_keywords(latest_user_text)

        # 1. Load active canonical memory items from storage if conversation_id known
        canonical_items = []
        if conversation_id:
            try:
                with session_scope() as session:
                    canonical_items = list_memory_items(
                        session,
                        conversation_id,
                        status=MemoryStatus.ACTIVE.value,
                    )
            except Exception as exc:
                logger.warning("Failed loading canonical memory for %s: %s", conversation_id, exc)

        # 2. Check and compress oversized tool outputs if needed
        processed_messages: list[dict[str, Any]] = []
        for idx, m in enumerate(messages):
            if m.get("role") == "tool" and len(str(m.get("content") or "")) > 400:
                raw_content = str(m.get("content") or "")
                compressed = raw_content
                if memory_ai is not None:
                    try:
                        norm = normalize_message(m, ordinal=idx)
                        compressed = await compress_tool_message(norm, adapter=memory_ai)
                    except Exception:
                        compressed = raw_content
                elif len(raw_content) > 1000:
                    # Deterministic tool compaction for very large outputs
                    lines = raw_content.splitlines()
                    head = "\n".join(lines[:10])
                    compressed = f"{head}\n... [tool output truncated for context budget: {len(lines)} lines total]"
                new_m = dict(m)
                new_m["content"] = compressed
                processed_messages.append(new_m)
            else:
                processed_messages.append(m)

        # 3. Build selectable candidate items
        candidates = []
        total_msgs = len(processed_messages)

        # Message candidates
        for idx, m in enumerate(processed_messages):
            is_latest = (idx == total_msgs - 1)
            item = score_message_item(
                m,
                index=idx,
                total_messages=total_msgs,
                query_keywords=query_keywords,
                is_latest=is_latest,
            )
            candidates.append(item)

        # Canonical memory candidates
        for c_idx, c_item in enumerate(canonical_items):
            item = score_canonical_item(
                c_item,
                ordinal=100_000 + c_idx,  # Distinct ordinal space for memory items
                query_keywords=query_keywords,
            )
            candidates.append(item)

        # Fast path: if no canonical memory and original already fits within budget, return as-is
        current_tokens = estimate_messages_tokens(messages)
        if not canonical_items and current_tokens <= target_budget:
            version_num = None
            if conversation_id and persist_snapshot:
                version_num = _persist_context_snapshot(
                    conversation_id,
                    messages,
                    budget=target_budget,
                    total_tokens=current_tokens,
                    canonical_count=0,
                    source_message_ids=[str(m.get("id") or idx) for idx, m in enumerate(messages)],
                )
            return CompileResult(
                messages=messages,
                total_tokens=current_tokens,
                context_version=version_num,
                canonical_items_used=0,
                selected_count=len(messages),
                budget=target_budget,
            )

        # 4. Constrained selector optimization
        selected = select_items_for_budget(candidates, target_budget)

        # 5. Assemble context messages in proper structure
        compiled_messages = assemble_context_messages(
            selected,
            has_canonical_memory=bool(canonical_items),
        )

        final_tokens = estimate_messages_tokens(compiled_messages)
        canonical_used = sum(1 for s in selected if s.kind == "canonical_memory")

        # 6. Snapshot persistence
        version_num = None
        if conversation_id and persist_snapshot:
            source_ids = [s.item_id for s in selected]
            version_num = _persist_context_snapshot(
                conversation_id,
                compiled_messages,
                budget=target_budget,
                total_tokens=final_tokens,
                canonical_count=canonical_used,
                source_message_ids=source_ids,
            )

        return CompileResult(
            messages=compiled_messages,
            total_tokens=final_tokens,
            context_version=version_num,
            canonical_items_used=canonical_used,
            selected_count=len(selected),
            budget=target_budget,
        )

    except Exception as exc:
        logger.exception("Context compilation failed; failing open with original messages: %s", exc)
        return CompileResult(
            messages=messages,
            total_tokens=estimate_messages_tokens(messages),
            budget=target_budget,
        )

File: src/context/compiler.ts
import type { Database } from "../db";
import { assembleContextMessages } from "./assembler";
import {
  extractKeywords,
  scoreCanonicalItem,
  scoreMessageItem,
  selectItemsForBudget,
  type SelectableItem,
} from "./selector";
import { estimateMessagesTokens } from "./tokens";
import { compressToolMessage } from "../memory/compressor";
import { normalizeMessage } from "../memory/ids";
import {
  latestContextVersion,
  resolveActiveConflicts,
} from "../memory/state";
import { retrieveActiveMemories, expandRelations } from "../memory/retrieve";
import type { ShortTermContextStore } from "../memory/context-store";
import type { MemoryAIAdapter } from "../providers/memory-ai";
import { contextVersions } from "../db/schema/context";

export interface CompileResult {
  messages: Array<Record<string, any>>;
  totalTokens: number;
  contextVersion?: number | null;
  canonicalItemsUsed: number;
  shortTermItemsUsed: number;
  selectedCount: number;
  budget: number;
}

async function persistContextSnapshot(
  db: Database,
  userId: string,
  compiledMessages: Array<Record<string, any>>,
  options: {
    budget: number;
    totalTokens: number;
    canonicalCount: number;
    sourceMessageIds: string[];
  }
): Promise<number | null> {
  try {
    const nextVer = (await latestContextVersion(db, userId)) + 1;
    const stateData = {
      budget: options.budget,
      total_tokens: options.totalTokens,
      message_count: compiledMessages.length,
      canonical_items_count: options.canonicalCount,
    };
    await db.insert(contextVersions).values({
      userId,
      version: nextVer,
      stateJson: JSON.stringify(stateData),
      sourceMessageIdsJson: JSON.stringify(options.sourceMessageIds),
      createdAt: new Date().toISOString(),
    });
    return nextVer;
  } catch {
    return null;
  }
}

export async function compileContext(
  db: Database | null,
  messages: Array<Record<string, any>>,
  userId?: string | null,
  options?: {
    budget?: number;
    memoryAi?: MemoryAIAdapter | null;
    persistSnapshot?: boolean;
    contextStore?: ShortTermContextStore | null;
  }
): Promise<CompileResult> {
  const targetBudget = options?.budget || 8000;

  if (!messages || messages.length === 0) {
    return {
      messages: [],
      totalTokens: 0,
      budget: targetBudget,
      canonicalItemsUsed: 0,
      shortTermItemsUsed: 0,
      selectedCount: 0,
    };
  }

  try {
    let latestUserText = "";
    for (let i = messages.length - 1; i >= 0; i--) {
      if (messages[i].role === "user") {
        latestUserText = String(messages[i].content || "");
        break;
      }
    }
    const queryKeywords = extractKeywords(latestUserText);

    let canonicalItems: any[] = [];
    let retrievedActive: any[] = [];
    if (db && userId) {
      try {
        // Lexical retrieval: pull the active long-term set narrowed by the
        // query keywords (OR match against content/value/predicate/subject),
        // then expand first-class relationships (supersedes/contradicts/
        // related) so we surface connected history without loading the whole
        // dataset. We deliberately do NOT hard-filter by scope: AI-extracted
        // scopes are unreliable, so a hard scope filter causes false negatives.
        retrievedActive = await retrieveActiveMemories(db, userId, {
          keywords: queryKeywords.size > 0 ? [...queryKeywords].slice(0, 6) : undefined,
          limit: 40,
        });

        const expanded = await expandRelations(db, userId, retrievedActive, {
          depth: 1,
          activeOnly: true,
          limit: 20,
        });
        const merged = new Map<number, any>();
        for (const item of [...retrievedActive, ...expanded]) {
          merged.set(item.id, item);
        }
        canonicalItems = resolveActiveConflicts(Array.from(merged.values()));
      } catch {}
    }

    // Short-term context from Redis (or in-memory store) — "what's happening right now?"
    let shortTermText = "";
    let shortTermItemsUsed = 0;
    const contextStore = options?.contextStore;
    if (contextStore && userId) {
      try {
        const ctx = await contextStore.getAllContext(userId);
        if (ctx && ctx.length > 0) {
          const lines = ctx.map((e) => `• ${e.key}: ${e.value}`);
          shortTermText = `[Short-Term Context (current state)]\n${lines.join("\n")}`;
          shortTermItemsUsed = ctx.length;
        }
      } catch {}
    }

    const processedMessages: Array<Record<string, any>> = [];
    for (let idx = 0; idx < messages.length; idx++) {
      const m = messages[idx];
      const contentStr = String(m.content || "");
      if (m.role === "tool" && contentStr.length > 400) {
        let compressed = contentStr;
        if (options?.memoryAi) {
          try {
            const norm = normalizeMessage(m, idx);
            compressed = await compressToolMessage(norm, options.memoryAi);
          } catch {
            compressed = contentStr;
          }
        } else if (contentStr.length > 1000) {
          const lines = contentStr.split("\n");
          const head = lines.slice(0, 10).join("\n");
          compressed = `${head}\n... [tool output truncated for context budget: ${lines.length} lines total]`;
        }
        processedMessages.push({ ...m, content: compressed });
      } else {
        processedMessages.push(m);
      }
    }

    const candidates: SelectableItem[] = [];
    const totalMsgs = processedMessages.length;

    for (let idx = 0; idx < totalMsgs; idx++) {
      const isLatest = idx === totalMsgs - 1;
      const item = scoreMessageItem(processedMessages[idx], {
        index: idx,
        totalMessages: totalMsgs,
        queryKeywords,
        isLatest,
      });
      candidates.push(item);
    }

    for (let cIdx = 0; cIdx < canonicalItems.length; cIdx++) {
      const item = scoreCanonicalItem(canonicalItems[cIdx], {
        ordinal: 100000 + cIdx,
        queryKeywords,
      });
      candidates.push(item);
    }

    // Deterministic retrieval happened above (scope-first + relationship
    // expansion); `retrievedActive` holds the active long-term set.

    const currentTokens = estimateMessagesTokens(messages);
    if (retrievedActive.length === 0 && shortTermItemsUsed === 0 && currentTokens <= targetBudget) {
      let versionNum: number | null = null;
      if (db && userId && options?.persistSnapshot !== false) {
        versionNum = await persistContextSnapshot(db, userId, messages, {
          budget: targetBudget,
          totalTokens: currentTokens,
          canonicalCount: 0,
          sourceMessageIds: messages.map((m, idx) => String(m.id || idx)),
        });
      }
      return {
        messages,
        totalTokens: currentTokens,
        contextVersion: versionNum,
        canonicalItemsUsed: 0,
        shortTermItemsUsed: 0,
        selectedCount: messages.length,
        budget: targetBudget,
      };
    }

    const selected = selectItemsForBudget(candidates, targetBudget);

    // Prepend the short-term context block to the assembled system prompt.
    let compiledMessages = assembleContextMessages(selected, {
      hasCanonicalMemory: retrievedActive.length > 0,
    });
    if (shortTermText) {
      const sysIdx = compiledMessages.findIndex((m) => m.role === "system");
      if (sysIdx !== -1) {
        const existing = String(compiledMessages[sysIdx].content || "");
        compiledMessages[sysIdx].content = `${existing}\n\n${shortTermText}`.trim();
      } else {
        compiledMessages.unshift({ role: "system", content: shortTermText });
      }
    }

    const finalTokens = estimateMessagesTokens(compiledMessages);
    const canonicalUsed = selected.filter((s) => s.kind === "canonical_memory").length;

    let versionNum: number | null = null;
    if (db && userId && options?.persistSnapshot !== false) {
      const sourceIds = selected.map((s) => s.itemId);
      versionNum = await persistContextSnapshot(db, userId, compiledMessages, {
        budget: targetBudget,
        totalTokens: finalTokens,
        canonicalCount: canonicalUsed,
        sourceMessageIds: sourceIds,
      });
    }

    return {
      messages: compiledMessages,
      totalTokens: finalTokens,
      contextVersion: versionNum,
      canonicalItemsUsed: canonicalUsed,
      shortTermItemsUsed,
      selectedCount: selected.length,
      budget: targetBudget,
    };
  } catch {
    return {
      messages,
      totalTokens: estimateMessagesTokens(messages),
      budget: targetBudget,
      canonicalItemsUsed: 0,
      shortTermItemsUsed: 0,
      selectedCount: messages.length,
    };
  }
}
